return_inferencedata option for pm.sample

michaelosthege opened the pull request:
Most PyMC3 modeling workflows (should) at some point use ArviZ to save and load traces as `InferenceData` objects.

With a `sample(…, return_inferencedata=True/False)` option, the conversion may happen as early as possible, encouraging the user to use ArviZ from the start.

+ [x] wait for next ArviZ release to happen
+ [x] update PR such that
  + [x] pin to latest ArviZ version
  + [x] take out the `NotImplementedError`

+ [x] there are no breaking changes to the user-facing API - but a `FutureWarning` is added, allowing us to default to `return-inferencedata=True` in `v4.0.0` if it turns out to be a good idea
+ [x] in #3883, Oriol made some optimizations to avoid duplicate work of conversion. This PR does the same by doing the conversion already in `sample()` and passing the `idata` for convergence checks.
+ [x] are the changes—especially new features—covered by tests and docstrings? -- test was added, other tests remain untouched, so they assert that the user-facing API did not break
+ [x] consider adding/updating relevant example notebooks - I'll do that on a future PR
+ [x] right before it's ready to merge, mention the PR in the RELEASE-NOTES.md

michaelosthege committed:
mention arviz functions by name in warning

michaelosthege committed:
convert to InferenceData already in sample function
+ convert to InferenceData and save metadata to it already in sample()
+ pass idata instead of trace to convergence check, to avoid duplicate work
+ directly use arviz diagnostics instead of pymc3 aliases

michaelosthege committed:
fix refactoring bugs

michaelosthege committed:
fix indentation

michaelosthege commented:
This is the first stage - converting to `InferenceData` earlier.

I'm going to wait for the tests before adding the `return_inferencedata` option.

michaelosthege committed:
add return_inferencedata option
+ set to None
+ defaults to False

michaelosthege commented:
Tests are green.
Now pushing the addition of the `return_inferencedata` kwarg. The tests should remain green...

rpgoldman commented on pymc3/sampling.py:
```diff
@@ -387,6 +393,20 @@ def sample(
+    if return_inferencedata and not discard_tuned_samples:
+        raise NotImplementedError(
+            'arviz.from_pymc3 does not handle warmups yet. See ArviZ issue #1146.'
+            'For this reason, we can not return InferenceData tat includes warmup draws right now.'
```
There's a typo here : `tat [sic] includes ...`

rpgoldman replied:
Fixed in my edit commit.

rpgoldman commented on pymc3/sampling.py:
```diff
@@ -327,6 +329,10 @@ def sample(
         is drawn from.
 
         Sampling can be interrupted by throwing a ``KeyboardInterrupt`` in the callback.
+    return_inferencedata: bool
```
minor points:

1. the `return_inferencedata` should be `bool, optional`
2. this format for docstrings doesn't comply with the numpy format, which has spaces on both sides of the ":"

https://numpydoc.readthedocs.io/en/latest/format.html

rpgoldman replied:
Fixed in my edit commit.

michaelosthege replied:
thanks 👍

AlexAndorra replied:
I think I'm the one who deleted the first space before the ":" in a previous PR -- this was because when you double-click on an arg on the website, this annoyingly (and weirdly) selects the arg, the colon and the first word after the colon. 

When the first space is not there (like: `arg: details`), double-clicking only selects the arg, as it should. See [this page](https://arviz-devs.github.io/arviz/generated/arviz.plot_compare.html#arviz.plot_compare) of ArviZ doc for instance, for an example of both (double-click on `figsize` and then on `textsize`). 

So I'm in favor of keeping the previous version, if this doesn't have any negative side effects.

PS: I think that's because of this change that there is a merge conflict: Jon added some clarifications as to how step kwargs can be passed to step methods, and I just merged [his PR](https://github.com/pymc-devs/pymc3/pull/3908).

rpgoldman replied:
I'm afraid it *does* have a negative effect: Sphinx will not format correctly docstrings that don't have " : ". 

I had put in some of these ": " docstrings myself, but I'm afraid they are wrong.  See the numpy doc I linked above.

michaelosthege replied:
I have no opinion on this. What does our doc-master, @ColCarroll say?

rpgoldman replied:
I'll wait for @ColCarroll to weigh in, but I will point out that we agreed to use numpy style, and the numpy style description is quite clear about what it expects.

Here's an example of what Sphinx does wrong if  you don't put the spaces around the colons -- the variable names and types run together:

![image](https://user-images.githubusercontent.com/3274/80976610-443d8400-8de9-11ea-97c6-b450f76c4390.png)

To be honest, though, I can't figure out for sure when Sphinx formats properly and when it does the wrong thing.

rpgoldman committed:
Fix numpy docstring format.

Replaced "<varname>: <type>" with "<varname> : <type>" per numpy guidelines.

Fix spelling typo.

rpgoldman reviewed (approved):
Looks good to me, pending tests.

Probably needs some additional tests that exercise returning `InferenceData`.

michaelosthege committed:
pass model to from_pymc3 because of deprecation warning

michaelosthege committed:
add test for return_inferencedata option

michaelosthege committed:
Merge branch 'master' into return-inferencedata-option

michaelosthege commented:
I've added a test for the new kwarg.

As `.sample(` appears another 78 times in the tests, I'd propose to update those tests with the future PR that changes the default behavior.

So this PR is ready for review - I'll update the release notes when I address your upcoming feedback.

michaelosthege commented:
You'll notice a `NotImplementedError` when `return_inferencedata==True and discard_tuned_samples==False`.

I'm in contact with Oriol to get `arviz.from_pymc3` support for warmup draws. If this can be arranged in an ArviZ release in the near future, we can use that and pin that ArviZ version for the `3.9` release?

AlexAndorra reviewed (approved):
Thanks Michael, this looks good to me 👌 
I'll merge tomorrow morning, in case @rpgoldman or @ColCarroll request some last-minute changes.

michaelosthege committed:
advise against keeping warmup draws in a MultiTrace

michaelosthege committed:
mention #3911

michaelosthege committed:
Merge branch 'return-inferencedata-option' of https://github.com/michaelosthege/pymc3 into return-inferencedata-option

AlexAndorra commented on pymc3/sampling.py:
```diff
@@ -412,6 +419,30 @@ def sample(
+
+    if return_inferencedata is None:
+        warnings.warn(
+            "In v4.0.0, pm.sample will return an `arviz.InferenceData` object instead of a `MultiTrace` by default. "
```
Shouldn't we say "In v3.10", as v4.0 will probably be reserved for the first release of PyMC4, as pointed out by @ColCarroll ?

michaelosthege replied:
changed it yesterday to "a future release"

AlexAndorra commented on pymc3/sampling.py:
```diff
@@ -412,6 +419,30 @@ def sample(
+        raise NotImplementedError(
+            'arviz.from_pymc3 does not handle warmups yet. See ArviZ issue #1146.'
+            'For this reason, we can not return InferenceData that includes warmup draws right now.'
+        )
```
I think this warning should be deleted, as the ArviZ PR will be merged in the next few days -- and will definitely be merged when this goes out with 3.9

michaelosthege replied:
I will take this out as soon as the ArviZ release number is known - then we can pin on that, or check the version here to maintain compatibility with ArviZ 0.7.0, if anything speaks against pinning `arviz>=0.?.?`

AlexAndorra replied:
I'd be in favor of pinning latest ArviZ, as this would simplify the code and workflow, and 0.x versions aren't supposed to be "production stable". But I'm of course open to debate.
I'll let know as soon as the ArviZ number is known

michaelosthege replied:
Done!

AlexAndorra commented on pymc3/sampling.py:
```diff
@@ -562,14 +593,27 @@ def sample(
+        # save additional sampling metadata
+        idata.posterior.attrs['n_tune'] = n_tune
+        idata.posterior.attrs['n_draws'] = n_draws
+        idata.posterior.attrs['t_sampling'] = t_sampling
```
Isn't all that done under the hood by `arviz.from_pymc3` now (cf Oriol's PR)?

michaelosthege replied:
Yes it is. Same reason as above.. I'm looking forward to take it out.

AlexAndorra replied:
Indeed, this will be solved if we pin latest ArviZ 👌

michaelosthege replied:
The version is now pinned and the code got much simpler. Please check the latest changes.

OriolAbril commented on pymc3/sampling.py:
```diff
@@ -562,14 +593,27 @@ def sample(
+        # that's why here, we make sure to only put actual posterior samples into idata.posterior
+        trace_warmup = trace[:-n_draws]   # <-- may result in len(trace_warmup) == 0
+        trace_posterior = trace[-n_draws:]
+        idata = arviz.from_pymc3(trace_posterior, log_likelihood=False, model=model)
```
There should be an option to set `log_likelihood=True` in case the user wants to calculate loo/waic or whatever with the returned idata object. The hardcoded false made sense when the inference data generated was thrown away. Similar thing would go for tuning/warmup storage, but this should be solved once ArviZ PR warmup is merged

OriolAbril replied:
Another option would be defaulting to None and set a `data.save_log_likelihood` as ArviZ rcParam for users who want to avoid memory overhead due to log_likelihood data.

AlexAndorra replied:
+1 for defaulting to None 👍 
People will indeed need the possibility of having the log likelihood in the idata (this is even the default on ArviZ, to compute LOO/WAIC).
I think the easiest way forward would be to hold this PR until ArviZ warmup PR is merged. That way, the different code blocks above can be properly refactored. WDYT?

michaelosthege replied:
I've changed it such that it defaults to the ArviZ default or rcParams if set.

I'm resolving this thread - please comment on the latest changes.

AlexAndorra commented:
@michaelosthege, as discussed, @OriolAbril's warmup PR was merged on ArviZ repo, and the next ArviZ release will be 0.8.0, so I think we can reactivate work on this PR and address the comments we made above 👌

michaelosthege commented:
Yes, we can wait for the ArviZ release.
I'll address the feedback when the release is out.

fonnesbeck commented:
We should probably either convert one of the example notebooks to using this throughout, or create a new one. The `api_quickstart` is probably the place to put it.

AlexAndorra commented:
@michaelosthege, ArviZ 0.8.0 hase been released 🍾 
We can reactivate work on this PR and address the comments we made above 👌

michaelosthege committed:
pin to arviz 0.8.0 and address review feedback

michaelosthege committed:
rerun/update notebook to show inferencedata trace

michaelosthege committed:
Merge branch 'master' into return-inferencedata-option

OriolAbril commented on pymc3/sampling.py:
```diff
@@ -561,15 +587,27 @@ def sample(
+    if compute_convergence_checks or return_inferencedata:
+        if 'data.save_log_likelihood' in arviz.rcParams:
+            # if the user change the arviz default, use it
+            save_ll = arviz.rcparams['data.save_log_likelihood']
```
3 comments on this: 
1. It should be `arviz.rcParams` like in matplotlib
2. I'll open an issue to add this rcParam to ArviZ so it gets added, it currently does not exist
3. Once implemented this if will be taken care of by `from_pymc3` so ti won't be needed, and before being implemented it cannot be triggered. I think this can actually be removed

michaelosthege replied:
Okay, if I understand this correctly, the setting can't be set? Can you make a 0.8.1 release then?

michaelosthege replied:
With the `idata_kwargs` setting, this is now user-configurable anyways.

OriolAbril commented on pymc3/sampling.py:
```diff
@@ -561,15 +587,27 @@ def sample(
+        else:
+            # otherwise, use the arviz default, which is True as of v0.8.0
+            save_ll = True
+        idata = arviz.from_pymc3(trace, log_likelihood=save_ll, model=model)
```
how about using some `idata_kwargs` or something like this? This way users can change pass dims and coords to `from_pymc3` and even a `prior` dict result of `pm.sample_prior_predictive`

michaelosthege replied:
Good idea! I did this in the latest commit.

michaelosthege committed:
fix typo

michaelosthege committed:
make all from_pymc3 accessible to the user

michaelosthege committed:
Merge branch 'return-inferencedata-option' of https://github.com/michaelosthege/pymc3 into return-inferencedata-option

AlexAndorra commented on RELEASE-NOTES.md:
```diff
@@ -20,6 +21,9 @@
 
 ### Maintenance
 - Tuning results no longer leak into sequentially sampled `Metropolis` chains (see #3733 and #3796).
+- Deprecated `sd` in version 3.7 has been replaced by `sigma` now raises `DeprecationWarning` on using `sd` in continuous, mixed and timeseries distributions. (see #3837 and #3688).
```
Is the beginning of the sentence missing some words or punctuation?

michaelosthege replied:
this line was actually a duplicate. I changed the wording in the deprecations section

AlexAndorra commented on pymc3/sampling.py:
```diff
@@ -275,61 +278,65 @@ def sample(
         is drawn from.
 
         Sampling can be interrupted by throwing a ``KeyboardInterrupt`` in the callback.
+    return_inferencedata : bool, optional
```
add "default=False", as above

AlexAndorra commented on pymc3/sampling.py:
```diff
@@ -275,61 +278,65 @@ def sample(
         Sampling can be interrupted by throwing a ``KeyboardInterrupt`` in the callback.
+    return_inferencedata : bool, optional
+        Whether to return the trace as an `arviz.InferenceData` (True) object or a `MultiTrace` (False)
+        Defaults to `False`, but we'll switch to `True` in version 4.0.0.
```
"version 3.10", as 4.0 will most probably be for PyMC4

michaelosthege replied:
I removed the specific version number. Let's discuss upcoming version numbers in the chat

AlexAndorra reviewed (changes requested):
Thanks @michaelosthege , looks good 👌 
Just added some comments, and I agree with Oriol's. Almost there 😉 

PS: didn't review the NB yet -- will do soon.
PPS: the tests are failing because of [this issue](https://github.com/arviz-devs/arviz/issues/1202), [fixed on ArviZ master](https://github.com/arviz-devs/arviz/pull/1203) -- Oriol will do a bugfix-release by tomorrow.

michaelosthege committed:
remove duplicate entry, and wording

michaelosthege committed:
address review feedback

michaelosthege commented:
waiting for #3931 because of pipeline failures

AlexAndorra reviewed (changes requested):
Thanks @michaelosthege ! 
I'm finishing the review of the example NB -- I have some comments and changes to implement, but not that big. After these changes, I think it will be ready to merge 🍾 
I don't understand how this is related to #3931 though 🤔

AlexAndorra commented on docs/source/notebooks/api_quickstart.ipynb:
```diff
@@ -16,12 +16,13 @@
      "name": "stdout",
```
<ul><li>Can get rid of %matplotlib inline</li><li>Let's replace all seaborn calls by ArviZ ones?</li><li>Maybe organize imports alphabetically, as in other NBs we modified recently?</li></ul><pre class="ql-syntax" spellcheck="false">import arviz as az
import matplotlib.pyplot as plt
import numpy as np
import pymc3 as pm
import theano.tensor as tt
</pre><ul><li>Let's use ArviZ style ;) And in another cell below (this has to do with how MPL sets defaults):</li></ul><pre class="ql-syntax" spellcheck="false">%config InlineBackend.figure_format = 'retina'
az.style.use('arviz-darkgrid')
print('Running on PyMC3 v{}'.format(pm.__version__))
</pre><ul><li>You can look at the NBs in the <a href="https://github.com/pymc-devs/resources" target="_blank">resources repo</a> for an example of how we format NBs.</li></ul>
<br/>

 _Reply via <a href='https://app.reviewnb.com/pymc-devs/pymc3/pull/3911/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='RESOLVED' data-cellIndex='1'/>

AlexAndorra replied:
<p>Typo above x2: "<s>sepcified</s> specify a different..."</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/pymc-devs/pymc3/pull/3911/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='RESOLVED' data-cellIndex='40'/>

AlexAndorra replied:
<p>Can you add a few lines about what <code>return_inferencedata</code> does, to introduce the difference with the default behavior, and the advantages of using <code>InferenceData</code>? You can link to <a href="https://arviz-devs.github.io/arviz/notebooks/Introduction.html" target="_blank">ArviZ quickstart</a> at the end.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/pymc-devs/pymc3/pull/3911/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='RESOLVED' data-cellIndex='57'/>

AlexAndorra replied:
<p>Let's use ArviZ directly, as you already returned <code>InferenceData</code> ;)</p><p><code>az.plot_trace(idata);</code></p><p> </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/pymc-devs/pymc3/pull/3911/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='RESOLVED' data-cellIndex='72'/>

AlexAndorra replied:
<p>Maybe use <code>az.summary</code> instead? More used in practice and contains <code>r_hat</code>  <span class="ql-cursor">﻿</span></p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/pymc-devs/pymc3/pull/3911/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='RESOLVED' data-cellIndex='74'/>

AlexAndorra replied:
<p><code>az.plot_forest</code></p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/pymc-devs/pymc3/pull/3911/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='RESOLVED' data-cellIndex='76'/>

AlexAndorra replied:
<p><code>az.plot_posterior</code></p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/pymc-devs/pymc3/pull/3911/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='RESOLVED' data-cellIndex='78'/>

AlexAndorra replied:
<p><code>az.plot_energy</code></p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/pymc-devs/pymc3/pull/3911/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='RESOLVED' data-cellIndex='80'/>

AlexAndorra replied:
<p>You can use <code>az.plot_kde</code></p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/pymc-devs/pymc3/pull/3911/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='RESOLVED' data-cellIndex='90'/>

AlexAndorra replied:
<p>You can use <code>az.plot_dist</code></p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/pymc-devs/pymc3/pull/3911/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='RESOLVED' data-cellIndex='93'/>

AlexAndorra replied:
<p>This example would be even nicer if you use the <code>pm.Data</code>  container, to illustrate :)</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/pymc-devs/pymc3/pull/3911/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='RESOLVED' data-cellIndex='102'/>

AlexAndorra replied:
<ul><li>After this last cell, can you add a last cell using <a href="https://github.com/rasbt/watermark" target="_blank">the watermark package</a>:</li></ul><p> </p><pre class="ql-syntax" spellcheck="false">%load_ext watermark
%watermark -n -u -v -iv -w
</pre><ul><li>Finally, can you <a href="https://github.com/dfm/black_nbconvert" target="_blank">format the NB with Black</a>: <code>pip install black_nbconvert</code> and then <code>black_nbconvert /path/to/a/notebook.ipynb</code> on the command line -- this will reformat the NB inplace automatically 😉</li></ul>
<br/>

 _Reply via <a href='https://app.reviewnb.com/pymc-devs/pymc3/pull/3911/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='RESOLVED' data-cellIndex='105'/>

AlexAndorra replied:
<p>As per our discussion, this whole section should probably be replaced by something like:</p><blockquote>It's worth highlighting the design choice we made with&nbsp;<code>logp</code>.&nbsp;As you can see above,&nbsp;<code>logp</code>&nbsp;is being called with arguments, so it's a method of the model instance. More precisely, it puts together a function based on the current state of the model -- or on the state given as argument to&nbsp;<code>logp</code>&nbsp;(see example below).</blockquote><blockquote>For diverse reasons, we assume that a&nbsp;<code>Model</code>&nbsp;instance isn't static. If you need to use&nbsp;<code>logp</code>&nbsp;in an inner loop and it needs to be static, simply use something like&nbsp;<code>logp = model.logp</code>. Here is an example below -- note the caching effect and the speed up:</blockquote>
<br/>

 _Reply via <a href='https://app.reviewnb.com/pymc-devs/pymc3/pull/3911/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='RESOLVED' data-cellIndex='10'/>

AlexAndorra commented on docs/source/notebooks/api_quickstart.ipynb:
```diff
@@ -16,12 +16,13 @@
      "name": "stdout",
```
<p>Maybe get rid of <code>return_inferencedata</code> here, as it's False, and introduce it below, as you set it to True?</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/pymc-devs/pymc3/pull/3911/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='RESOLVED' data-cellIndex='44'/>

michaelosthege replied:
<p>I wanted it to be future-proof, given that with this PR, not passing it raises a FutureWarning...</p><p> </p><p>This example actually uses the <code>MultiTrace</code> API to access transformed variable samples.</p>

AlexAndorra replied:
Ah you're right, let's keep it to `False` explicitly then 👌

AlexAndorra commented on docs/source/notebooks/api_quickstart.ipynb:
```diff
@@ -16,12 +16,13 @@
      "name": "stdout",
```
<p>No need for "," between "Note" and "that".</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/pymc-devs/pymc3/pull/3911/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='RESOLVED' data-cellIndex='62'/>

michaelosthege replied:
<p>I'm removing this comment alltogether, since this is now reported by the logger anyways.</p>

AlexAndorra commented on docs/source/notebooks/api_quickstart.ipynb:
```diff
@@ -16,12 +16,13 @@
      "name": "stdout",
```
<p>You can get rid of the Theano warning with this at the top of the NB, with other imports:</p><pre class="ql-syntax" spellcheck="false">import warnings
warnings.simplefilter(action="ignore", category=FutureWarning)
</pre>
<br/>

 _Reply via <a href='https://app.reviewnb.com/pymc-devs/pymc3/pull/3911/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='RESOLVED' data-cellIndex='87'/>

michaelosthege replied:
<p>😂</p>

AlexAndorra commented on docs/source/notebooks/api_quickstart.ipynb:
```diff
@@ -16,12 +16,13 @@
      "name": "stdout",
```
<p>Nice touch using idata ;)</p><p> </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/pymc-devs/pymc3/pull/3911/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='RESOLVED' data-cellIndex='97'/>

OriolAbril replied:
<p>IIUC, this will only generate posterior predictive samples for the first 500 draws of the first chain. Could <code>samples</code>  be removed to use the default? </p>

michaelosthege replied:
<p>yes. typical example for a chain &amp; draw unbiased random thinning convenience method</p>

AlexAndorra commented on docs/source/notebooks/api_quickstart.ipynb:
```diff
@@ -16,12 +16,13 @@
      "name": "stdout",
```
<p>You can use <code>az.plot_ppc</code>  instead</p><p> </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/pymc-devs/pymc3/pull/3911/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='RESOLVED' data-cellIndex='100'/>

OriolAbril replied:
<p>This will require to add the post_pred dict as <code>posterior_predictive</code>  group in the inference data object. Try something like:</p><p> </p><pre class="ql-syntax" spellcheck="false">idata.concat(az.from_pymc3(posterior_predictive=post_pred), inplace=True)
az.plot_ppc(idata)
# this above will plot the kde for data, to also add the mean as above,
# the ax.axvline has to be kept
</pre><p> </p>

michaelosthege replied:
<p>At the moment this does not work - <code>pm.sample_posterior_predictive(idata.posterior)</code>  collapses the 3 chains, 1000 draws into 1 chain 3000 draws. <code>arviz.concat</code>  then refuses to concatenate because the dims don't match.</p>

OriolAbril replied:
does using `keep_size=True` in sample_posterior_predictive solve the issue?

michaelosthege replied:
I don'T know what I did wrong, but now it works

OriolAbril commented:
> I don't understand how this is related to #3931 though thinking

I am now finishing the patch release for ArviZ: 0.8.1. Until then, trace objects without report attribute trigger the bug and crash. In ArviZ 0.8.1 it won't crash (however, warmup handling won't behave correctly which is not too big a deal as it is what has happened always until now :/ )

Again, thanks @michaelosthege

OriolAbril commented:
ArviZ 0.8.1 released, it should be required instead of 0.8.0

michaelosthege committed:
update arviz to 0.8.1 because of bugfix

michaelosthege committed:
incorporate review feedback
+ more direct use of ArviZ
+ some wording things

michaelosthege committed:
use arviz plot_ppc

michaelosthege commented:
@AlexAndorra and @OriolAbril thank you for the feedback on the notebook! they were really straightforward to apply.

AlexAndorra commented on RELEASE-NOTES.md:
```diff
@@ -20,6 +21,8 @@
 ### Maintenance
 - Tuning results no longer leak into sequentially sampled `Metropolis` chains (see #3733 and #3796).
+- We'll deprecate the `Text` and `SQLite` backends and the `save_trace`/`load_trace` functions, since this is now done with ArviZ. (see [#3902](https://github.com/pymc-devs/pymc3/pull/3902))
+- ArviZ `v0.8.1` is now the minimum required version
```
This should say 0.8.2

michaelosthege replied:
good catch!

michaelosthege committed:
also ignore Visual Studio cache

michaelosthege committed:
fix warmup saving logic and test

michaelosthege committed:
require latest ArviZ patch

AlexAndorra commented on pymc3/sampling.py:
```diff
@@ -412,6 +422,25 @@ def sample(
+            "To avoid complications with downstream analysis, you should slice them out:\n"
+            "trace_warmup = trace[: -trace.report.n_draws]\n"
+            "trace_posterior = trace[-trace.report.n_draws :]\n"
+            "Alternatively, you can use InferenceData: `pm.sample(..., return_inferencedata=True)`",
```
Actually, I'm thinking that we should remove the advice about slicing and just nudge users towards using ID, which takes care of warmup and posterior groups automatically, don't you think? 
This will avoid complications in downstream analysis _and_ help users get used to ID, which is probably gonna be the new default anyway. 
cc @OriolAbril

michaelosthege replied:
I made the change. However, I'd like to remind us that the warmup groups are still in a hidden attribute of the `InferenceData`

AlexAndorra reviewed (changes requested):
Thanks a lot @michaelosthege ! There was a small typo left, and I left a question about how to handle warmups 😉 Going to review the NB now.

michaelosthege committed:
change warning to nuget users towards InferenceData

michaelosthege committed:
update ArviZ minimum version

AlexAndorra commented:
@michaelosthege , finished reviewing the NB -- just a few tweaks to add / typos to fix and we'll be good to go 👌

michaelosthege commented:
The pipelines fails because of https://github.com/arviz-devs/arviz/issues/1210

I could make a workaround for the test case, but I also think having `draws=0` is a valid use case..

michaelosthege committed:
address review feedback

michaelosthege committed:
start showing the FutureWarning about return_inferencedata in minor release >=3.10

AlexAndorra reviewed (commented):
The NB is a marvel now, thanks a lot for your patience and perseverance on this @michaelosthege !
Waiting for ArviZ 0.8.3 to be released, and I'll merge if tests pass then -- I really, really hope no other edge case will pop up this time 😅 🤞

AlexAndorra commented:
Problems seem to be fixed on ArviZ master, but before releasing 0.8.3, we need to make sure this really works. To do that, the best would be that Travis tests against ArviZ master, not only against the latest release -- @canyon289 do you know if it's possible and how easy it would be?
Also pinging @twiecki and @ColCarroll, who could have valuable input.

OriolAbril commented:
Given that @michaelosthege has successfully run the tests locally, I'll merge tomorrow to make release 0.8.3 and hope for the best.

AlexAndorra commented:
@OriolAbril was emboldened by your successful local tests and released ArviZ 0.8.3 @michaelosthege, so you can push your last changes and I'll merge if tests pass 🤞

michaelosthege committed:
require arviz>=0.8.3 for latest bugfix

OriolAbril commented:
:tada:

AlexAndorra reviewed (approved):
We've done it, well done guys 👏 Quite a team 💪
